hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/Sources.java

```diff
@@ -870,4 +876,111 @@ public static JmsSourceBuilder jmsQueueBuilder(DistributedSupplier<ConnectionFac
+     * method, do your own testing.
+     * <p>
+     * {@code createOutputFn} gets the {@link ResultSet} and creates desired
+     * output object. The function is called for each row of the result set,
+     * user should not call {@link ResultSet#next()} or any other
+     * cursor-navigating functions.
+     * <p>
+     * Example: <pre>{@code
+     *     p.drawFrom(Sources.jdbc(
+     *         () -> {
+     *             try {
+     *                 return DriverManager.getConnection(DB_CONNECTION_URL);
+     *             } catch (SQLException e) {
+     *                 throw ExceptionUtil.rethrow(e);
+     *             }
+     *         },
+     *         (con, parallelism, index) -> {
+     *             try {
+     *                 return con.prepareStatement("select * from TABLE where mod(id, ?) = ?);
+     *                 stmt.setInt(1, parallelism);
+     *                 stmt.setInt(2, index);
+     *                 return stmt.executeQuery();
+     *             } catch (SQLException e) {
+     *                 throw ExceptionUtil.rethrow(e);
+     *             }
+     *         },
+     *         resultSet -> {
+     *             try {
+     *                 return new Person(resultSet.getInt(1), resultSet.getString(2));
+     *             } catch (SQLException e) {
+     *                 throw ExceptionUtil.rethrow(e);
+     *             }
+     *         }))
+     * }</pre>
+     * <p>
+     * The source does not save any state to snapshot. If the job is restarted,
+     * it will re-emit all entries.
+     * <p>
+     * Any {@code SQLException} will cause the job to fail.
+     * <p>
+     * The default local parallelism for this processor is 1.
+     *
+     * @param connectionSupplier creates the connection
+     * @param resultSetFn creates a {@link ResultSet} using the connection,
+     *                    total parallelism and index
+     * @param createOutputFn creates output objects from {@link ResultSet}
+     * @param <T> type of output objects
+     */
+    public static <T> BatchSource<T> jdbc(
+            @Nonnull DistributedSupplier<Connection> connectionSupplier,
+            @Nonnull ResultSetForPartitionFunction resultSetFn,
+            @Nonnull DistributedFunction<ResultSet, T> createOutputFn
+    ) {
+        return batchFromProcessor("jdbcSource",
+                SourceProcessors.readJdbcP(connectionSupplier, resultSetFn, createOutputFn));
+    }
+
+    /**
+     * Convenience for {@link Sources#jdbc(DistributedSupplier,
+     * ResultSetForPartitionFunction, DistributedFunction)}.
+     * A non-distributed, single-worker source which fetches the whole resultSet
+     * with a single query.
+     * <p>
+     * Example: <pre>{@code
+     *     p.drawFrom(Sources.jdbc(
+     *         DB_CONNECTION_URL,
+     *         "select ID, NAME from PERSON",
+     *         resultSet -> {
+     *             try {
+     *                 return new Person(resultSet.getInt(1), resultSet.getString(2));
```

Comment:
sounds like we need to use custom function that can throw SqlException, so you don't need to repeatedly deal with checked exceptions